src/server/game/Battlefield/Zones/BattlefieldTB.cpp

```diff
@@ -0,0 +1,839 @@
+            Towers.insert(go->GetGUID());
+
+    // Init Graveyards
+    SetGraveyardNumber(BATTLEFIELD_TB_GRAVEYARD_MAX);
+
+    // Graveyards
+    for (uint8 i = 0; i < BATTLEFIELD_TB_GRAVEYARD_MAX; i++)
+    {
+        BfGraveyard* graveyard = new BfGraveyard(this);
+
+        // When between games, the graveyard is controlled by the defending team
+        graveyard->Initialize(GetDefenderTeam(), TBGraveyards[i].gyid);
+
+        if (Creature* creature = SpawnCreature(TBGraveyards[i].spiritEntry[GetDefenderTeam()], TBGraveyards[i].x, TBGraveyards[i].y, TBGraveyards[i].z, TBGraveyards[i].o, GetDefenderTeam()))
+            graveyard->SetSpirit(creature, GetDefenderTeam());
+        
+        m_GraveyardList[i] = graveyard;
+    }
+
+    // Time warning vars
+    warnedFiveMinutes = false;
+    warnedTwoMinutes = false;
+    warnedOneMinute = false;
+    
+    UpdateNPCsAndGameObjects();
+
+    return true;
+}
+
+bool BattlefieldTB::Update(uint32 diff)
+{
+    bool m_return = Battlefield::Update(diff);
+
+    // Minutes till battle preparation warnings
+    if (GetState() == BATTLEFIELD_INACTIVE)
+    {
+        if (m_Timer <= 5 * MINUTE * IN_MILLISECONDS && !warnedFiveMinutes)
+        {
+            warnedFiveMinutes = true;
+            SendWarning(TB_TEXT_PREPARATIONS_IN_5_MIN);
+        }
+
+        if (m_Timer <= 2 * MINUTE * IN_MILLISECONDS && !warnedTwoMinutes)
+        {
+            warnedTwoMinutes = true;
+            SendWarning(TB_TEXT_PREPARATIONS_IN_2_MIN);
+        }
+
+        if (m_Timer <= 1 * MINUTE * IN_MILLISECONDS && !warnedOneMinute)
+        {
+            warnedFiveMinutes = true;
+            SendWarning(TB_TEXT_PREPARATIONS_IN_1_MIN);
+        }
+    }
+    
+    if (!updatedNPCAndObjects)
+    {
+        if (m_updateObjectsTimer <= diff)
+        {
+            UpdateNPCsAndGameObjects();
+            updatedNPCAndObjects = true;
+        }
+        else
+            m_updateObjectsTimer -= diff;
+    }
+
+    if (m_saveTimer <= diff)
+    {
+        if (!IsWarTime())
+            sWorld->setWorldState(TB_WS_TIME_NEXT_BATTLE, m_Timer);
+        m_saveTimer = 60 * IN_MILLISECONDS;
+    }
+    else
+        m_saveTimer -= diff;
+
+    return m_return;
+}
+
+void BattlefieldTB::OnPlayerEnterZone(Player* player)
+{
+    if (!m_isActive)
+        RemoveAurasFromPlayer(player);
+
+    SendInitWorldStatesTo(player);
+}
+
+void BattlefieldTB::OnPlayerLeaveZone(Player* player)
+{
+    if (!m_isActive)
+        RemoveAurasFromPlayer(player);
+}
+
+void BattlefieldTB::OnPlayerJoinWar(Player* player)
+{
+    RemoveAurasFromPlayer(player);
+
+    player->SetPvP(true);
+
+    // Bonus damage buff for attackers
+    if (player->GetTeam() == GetAttackerTeam())
+        player->CastCustomSpell(SPELL_TOWER_ATTACK_BONUS, SPELLVALUE_AURA_STACK, GetData(BATTLEFIELD_TB_DATA_TOWERS_DESTROYED), player, TRIGGERED_FULL_MASK);
```

Comment:
You should not cast this spell if GetData(BATTLEFIELD_TB_DATA_TOWERS_DESTROYED) == 0
SPELLVALUE_AURA_STACK with 0 value will not work as you would expect it here